### PR TITLE
[BEAM-5041] Java Fn SDK Harness use pTransform to track processed graph

### PR DESCRIPTION
**Pull request by angoenka**
Java Sdk Harness used pCollections to keep track of computed consumers [here](https://github.com/apache/beam/blob/ff95a82e461bd8319d9733be60e75992ba90cd7c/sdks/java/harness/src/main/java/org/apache/beam/fn/harness/control/ProcessBundleHandler.java#L158). This is incorrect as consumers are based on pTransforms so pTransforms should be used to keep track of computed consumers.

In case of Flatten, this creates an issue where pTransforms having same input as that to flatten are not executed. This causes 

https://github.com/apache/beam/blob/ff95a82e461bd8319d9733be60e75992ba90cd7c/sdks/java/core/src/test/java/org/apache/beam/sdk/transforms/FlattenTest.java#L316 to fail.
------------------------

Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

It will help us expedite review of your Pull Request if you tag someone (e.g. `@username`) to look at it.

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Apex | Dataflow | Flink | Gearpump | Samza | Spark
--- | --- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go_GradleBuild/lastCompletedBuild/) | --- | --- | --- | --- | --- | ---
Java | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_GradleBuild/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex_Gradle/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow_Gradle/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink_Gradle/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump_Gradle/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza_Gradle/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark_Gradle/lastCompletedBuild/)
Python | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python_Verify/lastCompletedBuild/) | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/) </br> [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | --- | --- | --- | ---

### CODE REVIEW

**Comment by angoenka** [comment]
Run Java PreCommit

**Comment by angoenka** [comment]
R: @bsidhom @ryan-williams @tweise

**Comment by ryan-williams** [comment]
I'm not familiar with this part of the code, but the change makes sense based on your explanation in the OSS Runners chat.

Slightly obligatory: is there a test that could be added to verify the new/correct behavior?

File: sdks/java/harness/src/main/java/org/apache/beam/fn/harness/control/ProcessBundleHandler.java [hunk 1]
```diff
@@ -185,23 +184,26 @@ private void createRunnerAndConsumersForPTransformRecursively(
...
+              options,
+              beamFnDataClient,
+              beamFnStateClient,
+              pTransformId,
+              pTransform,
+              processBundleInstructionId,
+              processBundleDescriptor.getPcollectionsMap(),
+              processBundleDescriptor.getCodersMap(),
+              processBundleDescriptor.getWindowingStrategiesMap(),
+              pCollectionIdsToConsumers,
```
Comment (bsidhom) [inline sdks/java/harness/src/main/java/org/apache/beam/fn/harness/control/ProcessBundleHandler.java]: How/when was `pCollectionIdsToConsumers` modified previously? I would like to know whether this is the right place to update `processedPTransformIds`.

Comment (angoenka) [reply sdks/java/harness/src/main/java/org/apache/beam/fn/harness/control/ProcessBundleHandler.java]: pCollectionIdsToConsumers is consumed at the time of bundle execution to lookup consumer for the pCollection.
https://github.com/apache/beam/blob/45ea59f3ec18989ec59901fd29a717b7de2d8ca4/sdks/java/harness/src/main/java/org/apache/beam/fn/harness/DoFnPTransformRunnerFactory.java#L282

**Comment by angoenka** [comment]
There are no unit tests specific to ProcessBundleHandler.java
Most of the functionality is tested using VR tests at the moment.
We should add it but I don't want to do it in this PR if adding unit tests is not trivial.
I can file a jira to track it.

File: sdks/java/harness/src/main/java/org/apache/beam/fn/harness/control/ProcessBundleHandler.java (hunk 1 above)
Comment (bsidhom) [reply sdks/java/harness/src/main/java/org/apache/beam/fn/harness/control/ProcessBundleHandler.java]: Sorry, I meant when were elements added?

**Comment by lukecwik** [comment]
This change looks good to me. Thanks for figuring out the flaw in the existing logic since it assumed consumers were only added to the current PCollection that was being iterated over.

Ben, IMO the validates runner test is a better was to make sure that end to end this isn't broken across runners.

I looked at the Python SDK and it builds the transforms by using the topological height of the transforms. We could do something very similar and rely on `QueryablePipeline#getTopologicallyOrderedTransforms` instead of the current recursive descent logic. This would be a good follow up change if your interested.

**Comment by lukecwik** [comment]
Run Java PreCommit

**Comment by angoenka** [comment]
Yes, Topological sort is a better way to tackle this issue in more logical manner. Current implementation is very close to what topological sort would do but as we already have means to do topological sort, we should use `QueryablePipeline#getTopologicallyOrderedTransforms`
Jira to track this https://issues.apache.org/jira/browse/BEAM-5090

**Comment by lukecwik** [comment]
Run Java PreCommit

**Pull request merged**